### DISCUSSION TITLE
client.scatter a pandas DataFrane

### DISCUSSION THREAD

**Discussion by YarShev** [dask/distributed#7476] [General]
Hi there,

Since client.scatter distributes data in a round-robin fashion grouping by number of cores, I wonder how a pandas DataFrame is split up between workers. Suppose I have 3 workers and the following script.

```python
from distributed.client import Client
import pandas

client = Client()
df = pandas.DataFrame([[1, 2, 3], [4, 5, 6], [7, 8, 9]])

client.scatter(df)
```

Could you tell me how the df is split up between 3 workers?

Thanks in advance!

**Comment by fjetter** [comment]
It's not split up. We treat data that is passed to scatter as opaque python objects and scatter the objects as is.
If you want to split up a pandas dataframe, I suggest to look at the `dask.DataFrame` API, specifically [`from_pandas`](https://docs.dask.org/en/stable/generated/dask.dataframe.from_pandas.html)

**Reply by YarShev** [reply]
Why does the docs say that the list `range(10)` is scattered out in a round-robin fashion here https://distributed.dask.org/en/stable/locality.html#data-scatter? Is it because of the list is a range? If I passed a normal python list, would it be distributed to only one worker?

Aside: your `from_pandas` link doesn't work.

**Reply by fjetter** [reply]
Updated the link.

Indeed `range(10)` is distributed and a plain list would be as well. every python object passed to scatter will be treated as a unique object (i.e. it will have a dask key) and we'll send it to a worker. if there are multiple objects we try to send every object to another worker via round robin

in the case of `range(10)` or `list(range(10))` there are 10 object to distribute. In your case, there is only one, the dataframe.

**Reply by YarShev** [reply]
Got it! Thanks for your responses.

**Reply by YarShev** [reply]
@fjetter, btw, is it possible to get futures to respective partitions of the Dask DataFrames, after dd.from_pandas() is called?

**Reply by jrbourbeau** [reply]
You're looking for [`distributed.futures_of`](https://distributed.dask.org/en/stable/api.html#distributed.futures_of)

**Reply by YarShev** [reply]
@jrbourbeau, I tried to use `future_of` but it prints an empty list. Could you please tell me why?

```python
import distributed
from distributed.client import Client
import dask.dataframe as dd
import pandas

client = Client(n_workers=8, threads_per_worker=1)
df = pandas.DataFrame([1, 2])
ddf = dd.from_pandas(df, npartitions=2)
distributed.futures_of(ddf)
[]
```

**Reply by jrbourbeau** [reply]
I missed that you weren't persisting the Dask DataFrame to the cluster. You'll want to do that before the `futures_of` call so there are futures to retrieve 

```python
In [1]: import distributed
   ...: from distributed.client import Client
   ...: import dask.dataframe as dd
   ...: import pandas
   ...:
   ...: client = Client(n_workers=8, threads_per_worker=1)
   ...: df = pandas.DataFrame([1, 2])
   ...: ddf = dd.from_pandas(df, npartitions=2)

In [2]: ddf = ddf.persist()

In [3]: distributed.futures_of(ddf)
Out[3]:
[<Future: finished, type: pandas.core.frame.DataFrame, key: ('from_pandas-b9b1623f8e6b37578c24437f54aedc1e', 0)>,
 <Future: finished, type: pandas.core.frame.DataFrame, key: ('from_pandas-b9b1623f8e6b37578c24437f54aedc1e', 1)>]
```

**Reply by YarShev** [reply]
It works now. Thanks! I wonder why both the partitions are saved in the same worker? I thought partitions of a Dask DataFrame would be distributed in a round-robin fashion between workers.

```python
fs = distributed.futures_of(ddf)
client.who_has(fs)
{"('from_pandas-4baea7a71ed57934d2b90c0100c6f228', 0)": ('tcp://127.0.0.1:35009',),
 "('from_pandas-4baea7a71ed57934d2b90c0100c6f228', 1)": ('tcp://127.0.0.1:35609',)}
```

**Reply by YarShev** [reply]
@jrbourbeau, @fjetter, just a friendly reminder on my question above.

**Reply by YarShev** [reply]
cc @martindurant

**Reply by martindurant** [reply]
The do not look like the same worker to me!!

**Reply by YarShev** [reply]
@martindurant, sorry, my bad. Did read incorrectly.
